### PR TITLE
Openvms fixes

### PR DESCRIPTION
**Pull request by joukj**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by joukj** [comment]
This pull request tackels 2 problems. 

1) update install instruction to call "perl Configure" indirectly via @CONFIG. This makes sure the correct machine is selected. Calling "perl Configure" directly results in many build errors.

2) 10-main.config : fixed use of bvoth cflags and CFLAGS resulting in wrong compiler flags for specific pointersize.

**Comment by t8m** [comment]
ping @levitte

**Comment by levitte** [comment]
The change to `perl Configure` was deliberate, see 16b0e0fcb32f553b9302db997045950c6af5f805.  This ended up in OpenSSL 3.0 and newer.

However, you're right that we haven't done anything to help indicate desired pointer-size variants.  In that case, the simplest is to specify the config targets available to do so, like this:

``` console
$ perl Configure vms-alpha-p64 ...
```

What is true, though, is that the pointersize variants have been [ahem] forgotten for a while, so I'd expect them to cause build errors.  On the other hand, we're running default builds (`perl Configure` with no extra arguments, `mms` and `mms test`) on buildbot and they build fine most of the time.  I'm curious to know what you stumbled into.  Care to show us?

**Comment by levitte** [comment]
Feel free to have a look at our buildbot builds: https://ci.buildbot.openssl.org/#/builders
The VMS builds should be easy to find, they have "vms" in their names.

**Comment by joukj** [comment]
I do not know what went wrong all the previous times with perl Configure but now it works. I updated the install.md accordingly. I added the specific pointer options for OpenVMS.

When applying the fix in the 10-main.conf file, the 64-bit version can be compiled. It passes all tests.

File: INSTALL.md [hunk 1]
```diff
@@ -1259,7 +1259,19 @@ the same.
...
+    should use one of the following options:
+    on AXP:
+         vms-alpha  (default option)
+         vms-alpha-p32
+         vms-alpha-p64
+    on IA64:
+         vms-ia64 (default option)
+         vms-ia64-p32
+         vms-ia64-p64
+    on X86: not yet implemented)
```
Comment (levitte) [inline INSTALL.md:1274]: I would place this a little further down, in the "Manual Configuration" section seen below, or in NOTES-VMS.md.

Comment (joukj) [reply INSTALL.md:1274]: OK to me

File: INSTALL.md [hunk 2]
```diff
@@ -1330,7 +1342,7 @@ directory and invoking the configuration commands from there.
     $ set default sys$login:
     $ create/dir [.tmp.openssl-build]
     $ set default [.tmp.openssl-build]
-    $ perl D:[PATH.TO.OPENSSL.SOURCE]Configure [options...]
+    $ @D:[PATH.TO.OPENSSL.SOURCE]Config [options...]
```
Comment (levitte) [inline INSTALL.md:1345]: Please undo this too

Comment (joukj) [reply INSTALL.md:1345]: sure I missed this one

**Comment by levitte** [comment]
I'm curious what prompted the "cflags" → "CFLAGS" change.  You mentioned a conflict, would you mind telling me more about it?

The idea with the casing is that "cflags" are mandatory, while "CFLAGS" are modifiable by the user like this:

``` console
$ perl Configure "CFLAGS=/whatever"
```

But if you chose the pointer size specific config target, what could you do that would conflict?  I'm assuming something like this:

``` console
$ perl Configure vms-ia64-p64 "CFLAGS=/whatever"
```

The idea here is that the `cflags` attribute contents would be combined with `/whatever`.  With the attribute `CFLAGS`, the attribute contents would be *replaced* by `/whatever`, thus rendering the idea of using `vms-ia64-p64` perfectly useless.

File: INSTALL.md [hunk 3]
```diff
@@ -1330,7 +1342,7 @@ directory and invoking the configuration commands from there.
     $ set default sys$login:
     $ create/dir [.tmp.openssl-build]
     $ set default [.tmp.openssl-build]
-    $ perl D:[PATH.TO.OPENSSL.SOURCE]Configure [options...]
+    $ @ perl D:[PATH.TO.OPENSSL.SOURCE]Configure [options...]
```
Comment (levitte) [inline INSTALL.md:1345]: `@`?

**Comment by joukj** [comment]
The problem is that if  the you have 
    cflags=>add("/pointer=64=args")
the qualifier does not end up in de CFLAGS macro in the resulting descrip.mms file.

**Comment by levitte** [comment]
> The problem is that if the you have cflags=>add("/pointer=64=args") the qualifier does not end up in de CFLAGS macro in the resulting descrip.mms file.

... oh!  Errrrrr, that's a bug in the descrip.mms template, then.  Thanks for the heads up, I'll have to look more closely at that.

**Comment by levitte** [comment]
> > The problem is that if the you have cflags=>add("/pointer=64=args") the qualifier does not end up in de CFLAGS macro in the resulting descrip.mms file.
> 
> ... oh! Errrrrr, that's a bug in the descrip.mms template, then. Thanks for the heads up, I'll have to look more closely at that.

I take that back.  The `CFLAGS` variable is still user-modifiable, so we're not actually putting the mandatory stuff (i.e. what comes from the `cflags` config attribute) there.  Instead, they get hard-coded into the compilation command lines themselves.  If you do something like this after having configured with the pointer_size variants, I believe you'll have a screenful and more of results:

```
$ SEARCH descrip.mms POINTER_SIZE
```

**Comment by joukj** [comment]
Right, but not all sources are compiled with these lines. For the 64-bit build you will end up with a mixture of 64-bit and 32-bit pointers resulting in failed tests and program crashes.

I see the default .c.obj rule is not redefined. Probably defining this rule with the compiling command now only inserted for a few sources

**Comment by joukj** [comment]
In the descrip.mms file the /pointer_size appears 57 times. So only 57 sources are compiled with 64-bit pointer sizes.
In total there are 1563 .c files in 121 directories......

**Comment by levitte** [comment]
Oh, right you are!  And I believe I've found the bugs that cause this

**Comment by levitte** [comment]
> I see the default .c.obj rule is not redefined. Probably defining this rule with the compiling command now only inserted for a few sources

We don't use the default rules, at all.  This is due to each compilation being tailored specifically for that compilation.  Not all compilations get exactly the same defines, for example.

**Comment by levitte** [comment]
Now that #20737 went in, were you able to build more consistently, @joukj?  That should make this PR be about just INSTALL.md at this point, no?

**Comment by levitte** [comment]
Side note: with #20737, I tried the `vms-ia64-p32` and `vms-ia64-p64` config targets.  The former builds and tests fine, while the latter...  has issues.  I'm looking into it.

**Comment by nhorman** [comment]
@joukj from the last comment about a year ago, it sounds like this or needs to do less than what it's currently doing, but we've not heard back from you.  Given the latency here, I'm closing this.  Please reopen if you would like to pick this back up

**Pull request closed without merging**